File: 프로그래머스/2/172927. 광물 캐기/광물 캐기.java
import java.util.*;

class Solution {
    static int answer; 
    static int[] pickArr;
    static int pickCnt;
    static String[] mineralArr;
    
    public int solution(int[] picks, String[] minerals) {
        answer = Integer.MAX_VALUE;
        
        pickArr = picks;
        pickCnt = 0;
        for(int i = 0; i < 3; i++){
            pickCnt += picks[i];
        }
        mineralArr = minerals;
        
        dfs(0, 0);
        return answer;
    }
    
    public static void dfs(int count, int point){
        // System.out.println(pickCnt + " " + count + " " + point);
        if(pickCnt == count || count * 5 > mineralArr.length){
            answer = Math.min(answer, point);
            return;
        }
        
        if(point > answer){
            return;
        }
        for(int i = 0; i < 3; i++){
            if(pickArr[i] > 0){
                pickArr[i]--;
                int getPoint = countingPoint(count, i);
                point += getPoint;
                dfs(count+1, point);
                point -= getPoint;
                pickArr[i]++;
            }
        }
    }
    public static int countingPoint(int cnt, int kind){
        int point = 0;
        for(int i = cnt*5; i < mineralArr.length && i < (cnt+1) * 5; i++){
            switch (kind) {
                case 0: 
                    point +=1;
                    break;
                case 1: 
                    point += mineralArr[i].equals("diamond") ? 5 : 1;
                    break;
                case 2:
                    point += mineralArr[i].equals("diamond") ? 25 : (mineralArr[i].equals("iron") ? 5 : 1);
                    break;
            }
        }
        
        return point;
    }
}

File: 프로그래머스/2/172927. 광물 캐기/광물 캐기.js
let answer = Number.MAX_VALUE;
let mLen = 0;
function solution(picks, minerals) {
    const dia = picks[0];
    const iron = picks[1];
    const stone = picks[2];
    const pickCnt = picks.reduce((acc, e) => {
        return acc + e;
    }, 0)
    mLen = minerals.length;
    
    dfs(0, 0, dia, iron, stone, pickCnt, minerals);
    
    return answer;
}
const dfs = (cur, fatigue, d, i, s, pickCnt, minerals) => {
    if(pickCnt <= 0 || mLen <= cur){
        answer = Math.min(answer, fatigue);
        return;
    }
    
    if(d > 0){
        const temp = calc(cur, fatigue, 'D', minerals);
        dfs(temp[0], temp[1], d - 1, i, s, pickCnt - 1, minerals);
    }
    
    if(i > 0){
        const temp = calc(cur, fatigue, 'I', minerals);  
        dfs(temp[0], temp[1], d, i - 1, s, pickCnt - 1, minerals);
    }
    
    if(s > 0){
        const temp = calc(cur, fatigue, 'S', minerals);  
        dfs(temp[0], temp[1], d, i, s - 1, pickCnt - 1, minerals);
    }
}
const calc = (s, f, type, minerals) => {
    let e = s + 5 <= mLen ? s + 5 : mLen;
    
    for(s; s < e; s++){
        const m = minerals[s];
        
        if(type === 'D') f += 1;
        else if(type === 'I') {
            if(m === 'diamond') f += 5;
            else f+= 1
        }else {
            if(m === 'diamond')f += 25;
            else if(m === 'iron')f += 5;
            else f += 1;
        }
    }
    return [e, f];
}